expression/builtin_cast.go

```diff
@@ -958,6 +972,9 @@ func (b *builtinCastStringAsDurationSig) evalDuration(row []types.Datum) (res ty
 	res, err = types.ParseDuration(val, b.tp.Decimal)
 	if types.ErrTruncatedWrongVal.Equal(err) {
 		err = sc.HandleTruncate(err)
+		if res == types.ZeroDuration {
```

Comment:
ditto

Reply:
I think when `res == types.ZeroDuration`, it maybe is a really zero duration, because the res is a value copy of types.ZeroDuration, types.ParseDuration return duration value, not duration pointer, so It can not distinguish it. Maybe it is better to add a boolean isNull return value to ParseDuration?

Reply:
there are 3 cases:
1. with ErrTruncatedWrongVal and zero duration, set isNull to true, append warning or not, then will show <nil>.
2. with ErrTruncatedWrongVal and not zero duration, the over range case,  append warning or not.
3. zero duration without error, the normal case, just pass.

so I think res == types.ZeroDuration condition inside err check statment will be ok.

Reply:
we can trust the result of `types.ParseDuration(val, b.tp.Decimal)` only if `err == nil`, which means we should move `res == types.ZeroDuration` condition outside the err check statement, that is:

```go
res, err = types.ParseDuration(val, b.tp.Decimal)
if err != nil {
    if types.ErrTruncatedWrongVal.Equal(err) {
        ...
    }
    ...
} else if res == types.ZeroDuration {
    ...
}
```